src/model/entity/room.go

```diff
@@ -1,10 +1,82 @@
 package entity
 
+import (
+	"errors"
+	"math/rand"
+)
+
+type RoomType uint
+
+const (
+	RoomTypeStart RoomType = iota
+	RoomTypeOrdinary
+	RoomTypeFinish
+)
+
+const (
+	RoomMinHeight = 4
+	RoomMinWidth  = 4
+	RoomWalls     = 2
+)
+
+type TileType uint
+
 type Room struct {
 	Shape   Box
+	Type    RoomType
 	Foods   []Food
 	Elixirs []Elixir
 	Scrolls []Scroll
 	Weapons []Weapon
 	Enemies []Enemy
+	Portal  *Portal
+}
+
+func NewRoom(roomType RoomType, minX int, minY int, maxWidth int, maxHeight int) (*Room, error) {
+	if maxWidth < RoomMinHeight || maxHeight < RoomMinHeight {
+		return nil, errors.New("room cannot be built: map is too small to fit a room of minimum required size")
+	}
+
+	width, height, err := calculateRoomSize(maxWidth, maxHeight)
+	if err != nil {
+		return nil, err
+	}
+
+	shape := Box{
+		Point: Point2D[int]{X: minX + rand.Intn(int(width)), Y: minY + rand.Intn(int(height))},
+		Size:  Size2D[uint]{Height: height, Width: width},
+	}
+
+	var portal *Portal = nil
+	if roomType == RoomTypeFinish || roomType == RoomTypeStart {
+		portal = generatePortal(shape.Point.X, shape.Point.Y, int(width), int(height))
+	}
+
+	return &Room{
+		Shape:   shape,
+		Type:    roomType,
+		Foods:   []Food{},
+		Elixirs: []Elixir{},
+		Scrolls: []Scroll{},
+		Weapons: []Weapon{},
+		Enemies: []Enemy{},
+		Portal:  portal,
+	}, nil
+}
+
+func calculateRoomSize(maxWidth int, maxHeight int) (uint, uint, error) {
```

Comment:
```suggestion
func calculateRoomSize(maxSize Size2D[uint]) (Size2D[uint], error) {
```

Reply:
Ещё, для единообразия и более чёткого смысла в имени функции, можно заменить глагол `calculate` на `generate`. Ведь у нас именно генерация происходит на основе случайности, а не вычисление, которые всегда детерминировано на основе одних и тех же входных данных.

```suggestion
func generateRoomSize(maxWidth int, maxHeight int) (uint, uint, error) {
```

Reply:
Также можно передавать ограничители по min/max значениям в качестве аргументов. Это снизит уровень связанности кода, что хорошо.